### PR TITLE
#22789 Creating variant Database table and TU

### LINKED ISSUE #22789: Create Variants Concept into DotCMS
We need to support the `variants` concept into DotCMS so we are going to:

- [x] Create  a new `variants` table: id varchar(36), name varchar(255), archive. We must to update the posgres.sql and mssql.sql file to create this new table Also we need to create a new Upgrade Task
- [x] Create a `Variant` class with to attributes: id and name
- [x] Create `VariantAPI` with methods: save, update, delete, find by id and  archive
- [x] Create `VariantFactory` with methods: save, update, delete, find by id and archive
- [x] Use cache for Variant

### PR DESCRIPTION
**Pull request by freddyucv**
Creating variant Database table and TaskUpgrate

first step for https://github.com/dotcms/core/issues/22789

### CODE REVIEW

File: dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task222208CreateVariantTable.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import com.dotmarketing.db.DbConnectionFactory;
+import com.dotmarketing.exception.DotDataException;
+import com.dotmarketing.exception.DotRuntimeException;
+import com.dotmarketing.startup.StartupTask;
+import java.sql.SQLException;
+
+public class Task222208CreateVariantTable implements StartupTask {
+
+    private final String POSTGRES_QUERY = "CREATE TABLE IF NOT EXISTS variant ("
+            + "  id varchar(36) primary key,"
```
Comment (wezell) [inline dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task222208CreateVariantTable.java:15]: No reason to be limited to 36 chars. Make it 255 at least

**Review by wezell — CHANGES_REQUESTED** [review]

File: dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task222208CreateVariantTable.java (hunk 1 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task222208CreateVariantTable.java]: done https://github.com/dotCMS/core/pull/22827/commits/3939fd3672ff479e0d14a1d8639d8ec686640682

File: dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task222208CreateVariantTable.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import com.dotmarketing.common.db.DotDatabaseMetaData;
+import com.dotmarketing.db.DbConnectionFactory;
+import com.dotmarketing.exception.DotDataException;
+import com.dotmarketing.exception.DotRuntimeException;
+import com.dotmarketing.startup.StartupTask;
+import java.sql.SQLException;
+
+public class Task222208CreateVariantTable implements StartupTask {
+
+    private final String POSTGRES_QUERY = "CREATE TABLE IF NOT EXISTS variant ("
```
Comment (wezell) [inline dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task222208CreateVariantTable.java:14]: good work using `IF NOT EXISTS`. We need this type of logic on all of our startup tasks

File: dotCMS/src/main/resources/mssql.sql [hunk 3]
```diff
@@ -2684,3 +2684,9 @@ create table storage_x_data (
 
 CREATE TABLE shedlock(name VARCHAR(64) NOT NULL, lock_until datetime NOT NULL,
                       locked_at datetime NOT NULL, locked_by VARCHAR(255) NOT NULL, PRIMARY KEY (name));
+
+create table variant (
+     id NVARCHAR(255) primary key,
+     name NVARCHAR(255) not null,
+     deleted tinyint not null,
```
Comment (jdotcms) [inline dotCMS/src/main/resources/mssql.sql:2691]: this should be called archived and init it

File: dotCMS/src/main/resources/postgres.sql [hunk 4]
```diff
@@ -2483,5 +2483,9 @@ create table storage_x_data (
 CREATE TABLE shedlock(name VARCHAR(64) NOT NULL, lock_until timestamptz NOT NULL,
                       locked_at timestamptz NOT NULL, locked_by VARCHAR(255) NOT NULL, PRIMARY KEY (name));
 
-
+create table variant (
+     id varchar(255) primary key,
+     name varchar(255) not null,
+     deleted boolean NOT NULL default false
```
Comment (dsilvam) [inline dotCMS/src/main/resources/postgres.sql:2489]: Change to "archived"

**Review by dsilvam — CHANGES_REQUESTED** [review]

File: dotCMS/src/main/resources/mssql.sql (hunk 3 above)
Comment (freddyucv) [reply dotCMS/src/main/resources/mssql.sql]: done https://github.com/dotCMS/core/pull/22827/commits/9d53cd59ff07b5a9134e4d5df7d7741a7f0e4a60

File: dotCMS/src/main/resources/postgres.sql (hunk 4 above)
Comment (freddyucv) [reply dotCMS/src/main/resources/postgres.sql]: done https://github.com/dotCMS/core/pull/22827/commits/9d53cd59ff07b5a9134e4d5df7d7741a7f0e4a60

**Review by jdotcms — APPROVED** [review]

**Review by jdotcms — APPROVED** [review]

File: dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task222208CreateVariantTableTest.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+import java.sql.ResultSet;
+import java.sql.ResultSetMetaData;
+import java.sql.SQLException;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+public class Task222208CreateVariantTableTest {
```
Comment (dsilvam) [inline dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task222208CreateVariantTableTest.java:24]: Test needs to be added to MainSuite

**Review by dsilvam — CHANGES_REQUESTED** [review]

File: dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task222208CreateVariantTableTest.java (hunk 5 above)
Comment (freddyucv) [reply dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task222208CreateVariantTableTest.java:24]: done https://github.com/dotCMS/core/pull/22827/commits/280f4a175649315b2ac562270104d588598764d7

**Review by dsilvam — APPROVED** [review]

**Review by dsilvam — APPROVED** [review]

**Pull request merged**